### PR TITLE
fix health probes for VM's

### PR DESCRIPTION
**Pull request by rinormaloku**
This resolves the issue explained here: #34407

tl;dr 

VM's that specify the `ISTIO_SVC_IP` fail the health checks (this wasn't an issue in Istio 1.9.x when manually specifying the env variable)

With this PR we update the `istioctl workload entry configure` command to set the env var `REWRITE_PROBE_LEGACY_LOCALHOST_DESTINATION` to true, which will use localhost instead of IP address for the health probes.

### CODE REVIEW

**Comment by rinormaloku** [comment]
@stevenctl and @howardjohn, tagging you guys as you helped review the previous story regarding supporting VMs that specifying the workloadIP: #32383

**Comment by howardjohn** [comment]
I understand the problem but am not sure this is the correct fix. Will think about it more and get back to you soon

**Comment by howardjohn** [comment]
Specifically because this requires applications to bind to `localhost`, which historically has proven to be something many applications do not do

**Comment by howardjohn** [comment]
Thought about this some more:

One idea is to keep using external IP, but auto-detect the IP is external and not use the `::6` bind. This does not work; the request would loop back through envoy inbound and be rejected (strict mtls, etc).

Other option is to have a separate setting for internal IP and external IP. This would work but I think a bit complex.

Third option is to do exactly what you did here, but under a flag (say `--external-ip` or something) so it only impacts the (somewhat small) set of external IP users. For these users, they need to take care to ensure their applications bind to `localhost`.

I am leaning towards the third option, WDYT?

Other option is to find an alternative to "tag" traffic we don't want iptables captured besides source IP, but I have not found a good way

**Comment by rinormaloku** [comment]
For the third option would these be the changes: 
1. rename workloadIP flag to --external-ip (this appends the REWRITE_PROBE_LEGACY_LOCALHOST_DESTINATION env)
2. add a new --internal-ip flag that adds only the ISTIO_SVC_IP

Or only the first point?

**Comment by howardjohn** [comment]
> For the third option would these be the changes:
> 
> 1. rename workloadIP flag to --external-ip (this appends the REWRITE_PROBE_LEGACY_LOCALHOST_DESTINATION env)
> 2. add a new --internal-ip flag that adds only the ISTIO_SVC_IP
> 
> Or only the first point?

In a vacuum those both sound good. But would be slightly worried about renaming existing flag? I suppose it could be an alias. I don't have a strong opinion here, will leave to @istio/wg-user-experience-maintainers (owners of this code) to decide on best flag names

**Comment by rinormaloku** [comment]
Most likely nobody uses the flag `workloadIP` as Istio 1.11 is the first release containing it — and if someone would have used it they would have filed a ticket as the feature is not working. I think it is safe to rename those. However, I'll wait for a confirmation before I make and submit the updates

**Comment by howardjohn** [comment]
Oh I was not aware that it's new! No wonder no one else had this issue :-).

In that case, I think `--internal-ip` and `--external-ip` are good choices.

**Comment by rinormaloku** [comment]
@howardjohn made and tested the changes. 


Should I update this release note? https://github.com/istio/istio/blob/master/releasenotes/notes/32462.yaml

Or should I create a new one?

File: istioctl/cmd/workload.go
Comment (howardjohn) [inline istioctl/cmd/workload.go:372]: should add some validation that both external and internal are not set. Otherwise one is silently ignored.

**Review by howardjohn — APPROVED** [review]
LGTM aside from one small comment. Would be great to update existing release note, as it is now no longer accurate

**Review by esnible — CHANGES_REQUESTED** [review]
This seems reasonable to me.  I do not use the `workload` commands, so I am not sure.

You are removing the `workloadIP` parameter, replacing it with `internalIP` and `externalIP`.  Please keep `workloadIP`, but mark it as deprecated.  If it is supplied, print a warning then either fail the command or use it to set `internalIP`.

**Comment by howardjohn** [comment]
> You are removing the workloadIP parameter, replacing it with internalIP and externalIP. Please keep workloadIP, but mark it as deprecated. If it is supplied, print a warning then either fail the command or use it to set internalIP.


The flag is new in 1.11, so we never released it. That is why we should make sure to merge this in 1.11 (or not at all, in the current state)

**Review by esnible — APPROVED** [review]

**Review by shamsher31 — APPROVED** [review]
Please add a release notes.

File: istioctl/cmd/workload.go
Comment (rinormaloku) [reply istioctl/cmd/workload.go:372]: It is okay if both are not set, then we won't create a sidecar env configuration (sidecar.env) and by default, it is set by the istio-start.sh script:
```
ISTIO_SVC_IP=$(hostname --all-ip-addresses | cut -d ' ' -f 1)
```

**Comment by rinormaloku** [comment]
btw @howardjohn thinking about this, does it make more sense to have these flags:

--workloadIP and
--probe-on-localhost
?

I.e. workloadIP configures the IP that the workload uses to auto register and for health checks,

Meanwhile --probe-on-localhost overwrites so that the health check is performed on localhost for apps bound to it. (I think the --internal-ip and --external-ip might confuse users)

**Comment by howardjohn** [comment]
> Meanwhile --probe-on-localhost overwrites so that the health check is performed on localhost for apps bound to it. (I think the --internal-ip and --external-ip might confuse users)

idk, personally I think `--probe-on-localhost` is more confusing to users (then again, I am not a user :slightly_smiling_face: ). It specifies what we want to do, but a user needs to know *why* they want to do that - I doubt anyone will know that unless they get a broken VM and stumble upon this PR/issue?

Whereas with external/internal, they will see both flags and realize they need to decide which their IP is. Under the hood we make it work - which could be through different means in the future.

**Comment by rinormaloku** [comment]
Makes sense...

So, I am asking as I don't know the implementation details and what I am
not sure is:

Do we want to always set  REWRITE_PROBE_LEGACY_LOCALHOST_DESTINATION to
true when we have an external IP?

If yes then the current implementation is good :))

**Comment by howardjohn** [comment]
Today we do. Maybe in the future we have a better solution

**Comment by rinormaloku** [comment]
@howardjohn do we want to `/cherrypick release-1.11` this change?

**Comment by howardjohn** [comment]
hm, it was supposed to.. from the label

/cherrypick release-1.11

**Pull request merged**
